marconi-chain-index/src/Marconi/ChainIndex/Transformer/WithSyncLog.hs

```diff
@@ -0,0 +1,250 @@
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Marconi.ChainIndex.Transformer.WithSyncLog (
+  WithSyncStats (..),
+  LastSyncLog (..),
+  LastSyncStats (..),
+  withSyncStats,
+) where
+
+import Cardano.Api qualified as C
+import Cardano.BM.Trace (logInfo)
+import Control.Lens (makeLenses, (^.))
+import Control.Monad (when)
+import Control.Monad.Except (MonadError)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Data.IORef (IORef, modifyIORef', newIORef, readIORef)
+import Data.Maybe (isJust)
+import Data.Time (defaultTimeLocale, formatTime)
+import Data.Time.Clock (NominalDiffTime, UTCTime, diffUTCTime, getCurrentTime)
+import Data.Word (Word64)
+import GHC.Generics (Generic)
+import Marconi.ChainIndex.Orphans ()
+import Marconi.ChainIndex.Types (MarconiTrace, TipAndBlock (TipAndBlock))
+import Marconi.Core qualified as Core
+import Marconi.Core.Class (
+  Closeable,
+  IsIndex (index, rollback, setLastStablePoint),
+  IsSync,
+  Queryable,
+ )
+import Marconi.Core.Transformer.Class (IndexerTrans, unwrap)
+import Marconi.Core.Transformer.IndexTransformer (
+  IndexTransformer (IndexTransformer),
+  indexVia,
+  rollbackVia,
+  setLastStablePointVia,
+  wrappedIndexer,
+  wrapperConfig,
+ )
+import Prettyprinter (Pretty, pretty, (<+>))
+import Text.Printf (printf)
+
+-- | Chain synchronisation statistics measured starting from previously measured 'LastSyncStats'.
+data LastSyncStats = LastSyncStats
+  { syncStatsNumBlocks :: !Word64
+  -- ^ Number of applied blocks since last message
+  , syncStatsNumRollbacks :: !Word64
+  -- ^ Number of rollbacks since last message
+  , syncStatsChainSyncPoint :: C.ChainPoint
+  -- ^ Chain index syncing point
+  , syncStatsNodeTip :: C.ChainTip
+  -- ^ Current node tip
+  , syncStatsLastMessageTime :: !(Maybe UTCTime)
+  -- ^ Timestamp of last printed message
+  }
+  deriving (Eq, Show)
+
+-- | Logging datatype for information that occured since the previous 'LastSyncLog'.
+data LastSyncLog = LastSyncLog
+  { syncStatsSyncLog :: LastSyncStats
+  -- ^ Stats since the last syncing log message
+  , timeSinceLastMsgSyncLog :: Maybe NominalDiffTime
+  -- ^ Time since last syncing log message.
+  }
+  deriving stock (Eq, Show, Generic)
+
+-- | Logging datatype for information that occured since the previous 'LastSyncLog'.
+data SyncLog event = SyncLog
+  { _syncLogStats :: IORef LastSyncStats
```

Comment:
Furthermore, if you do so, you could introduce lenses for `SyncLog` and `LastSyncStats` and use them to do the state update (it may lead to cleaner update functions)

Reply:
Yep, my bad, I added the IORef right at the start and didn't revisit it. I'll change it.